### ISSUE TITLE
WASI Console use causes panic (haskell)

### ISSUE THREAD

**Issue by iluminae** [tetratelabs/wazero#1232]
**Describe the bug**
Haskell compiler wasm32-wasi-ghc version 9.7.20230306 compiles a WASM artifact that cannot use console (or maybe all) file descriptors. (see addtl. context below, I know this is not supported)

**To Reproduce**
using haskell code:

```haskell
module Query where

import System.IO (hPutStr,hFlush,stdout)

foreign export ccall query :: IO Int

query :: IO Int
query = do
  hPutStr stdout "{\"name\":\"bob\"}"
  hFlush stdout
  return 0
```

<details><summary>compiled with</summary>

```bash
wasm32-wasi-ghc Query.hs -o haskell.wasm -no-hs-main -optl-mexec-model=reactor -optl-Wl,--export=hs_init,--export=query
```
</details>

<details><summary>ran with</summary>

```go
ctx := context.Background()
r := wazero.NewRuntime(ctx)
defer r.Close(ctx)
wasi_snapshot_preview1.MustInstantiate(ctx, r)
mod, err := r.Instantiate(ctx, bin)
if err != nil { panic(err) }
// starts the haskell runtime
_, err = mod.ExportedFunction("hs_init").Call(ctx, 0, 0)
if err != nil { panic(err) }
// call our exported function
err = mod.ExportedFunction("query").Call(ctx)
if err != nil { panic(err) }
```
</details>

**Expected behavior**
the wasm binary should be able to write to stdout.

**Screenshots**
<details><summary>stack trace</summary>

```
panic: interface conversion: *sys.stdioFileReader is not io.Writer: missing method Write (recovered by wazero)
wasm stack trace:
        wasi_snapshot_preview1.poll_oneoff(i32,i32,i32,i32) i32
        .__wasi_poll_oneoff(i32,i32,i32,i32) i32
        .poll(i32,i32,i32) i32
        .fdReady(i32,i32,i64,i32) i32
        .base_GHCziIOziFD_zdwwriteRawBufferPtr_entry() i32
        .StgRun(i32,i32) i32
        .scheduleWaitThread(i32,i32,i32)
        .rts_inCall(i32,i32,i32)
        .query() i32
```
</details>

**Environment (please complete the relevant information):**
- Go version: 1.20.1
- wazero Version: v1.0.0-rc.1 (reproduced on main@7e953d7 as well)
- wasm32-wasi-ghc Version: 9.7.20230306
- Host architecture: amd64
- Runtime mode: interpreter

**Additional context**
I know this is not a tested or supported wasm compiler and I do not expect much support here, I just wanted to provide a data point on the WASI implementation not meeting expectations of a runtime. When I checked out the stack trace this lead me to believe it was trying to use a nonblocking IO call and waiting on the fd event, but it was getting an unexpected error and failing.

Hope it helps in the future!

**Comment by evacchi** [comment]
hey this is super-interesting, can you share the wasm binary somewhere for convenience? :)

**Comment by iluminae** [comment]
yep, see attached 
[haskell.wasm.gz](https://github.com/tetratelabs/wazero/files/10961698/haskell.wasm.gz)

**Comment by evacchi** [comment]
cool, will look into that 🧐

**Comment by mathetake** [comment]
> panic: interface conversion: *sys.stdioFileReader is not io.Writer: missing method Write (recovered by wazero)

hah I see cc @codefromthecrypt

**Comment by codefromthecrypt** [comment]
first issue was that when a target file descriptor isn't read, it ends up zero, which is STDIN, so this hid some underlying issue and solved in #1237

**Comment by codefromthecrypt** [comment]
@evacchi here's where I got..

So, if you add a test like this, it will fail

```

func TestHaskell(t *testing.T) {
	ctx := context.Background()
	//// Set context to one that has an experimental listener that logs all host functions.
	//ctx = context.WithValue(context.Background(), experimental.FunctionListenerFactoryKey{},
	//	logging.NewHostLoggingListenerFactory(os.Stdout, logging.LogScopeAll))
	r := wazero.NewRuntime(ctx)
	defer r.Close(ctx)

	wasi_snapshot_preview1.MustInstantiate(ctx, r)

	bin, err := os.ReadFile("/Users/adrian/Downloads/haskell.wasm")
	require.NoError(t, err)

	mod, err := r.InstantiateWithConfig(ctx, bin, wazero.NewModuleConfig().
		WithStdin(os.Stdin).
		WithStdout(os.Stdout).
		WithStderr(os.Stderr))
	require.NoError(t, err)

	// starts the haskell runtime
	_, err = mod.ExportedFunction("hs_init").Call(ctx, 0, 0)
	require.NoError(t, err)

	// call our exported function
	_, err = mod.ExportedFunction("query").Call(ctx)
	require.NoError(t, err)
}
```

Also, logging seems misaligned probably due to stdio being not skipped properly

While [lightly documented](https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-event_fd_readwrite-struct), my guess is that there is an implicit behavior people use for inlining a poll, as otherwise, no idea how you can determine how many bytes are available for read or write.

I don't think this can make the current release candidate, but it might be viable by 1.0 next week, or at least better explained where logging doesn't crash. Good luck!

**Comment by evacchi** [comment]
ok so, here's what I found out so far.

For some reason we are (mistakenly?) reading FD 0 from the linear memory struct `event`. 
In fact, you can print `fd` here:

```go
func processFDEvent(mod api.Module, eventType byte, inBuf []byte) Errno {
   fd := le.Uint32(inBuf)
```

and you'll see it's always 0; you can also slam `1` in there to see that the test will pass (more or less, there is still some issue with flushing I think -- if I add a last println() at the end of the test case it passes. Go figure)

```go
func processFDEvent(mod api.Module, eventType byte, inBuf []byte) Errno {
   fd := uint32(1) //le.Uint32(inBuf)
```

now I don't know if this is related but `inBuff[4]` in that function is suspiciously =1 while the other offsets are 0, so if I add `+(4)` (the parens are just for emphasis):

			errno = processFDEvent(mod, eventType, inBuf[inOffset+8+4+(4):])
			
again, this works. It is looking like a sort of off-by-one error, but I could be completely mistaken.

I managed to run this against wasmtime https://github.com/evacchi/go-wasm-boilerplate and that works. I also rebuilt the example with a `main` so that we don't need the boilerplate for embedding (to my great surprise `./setup.sh` works flawlessly https://gitlab.haskell.org/ghc/ghc-wasm-meta -- but don't put the `source $HOME/.ghc-wasm/env` in your profile rc because it messes up with the default compiler and cgo gets crazy)

long story short, it is still broken, but now it's easier to test.


[main-hputstrln.wasm.gz](https://github.com/tetratelabs/wazero/files/10970924/main-hputstrln.wasm.gz)

```hs
module Main where

import System.IO (hPutStr,hFlush,stdout)

main = do
  hPutStr stdout "{\"name\":\"bob\"}"
  hFlush stdout
  return 0
```


    wazero run main-hputstrln.wasm

dies the same way as the original example

[main-putstrln.wasm.gz](https://github.com/tetratelabs/wazero/files/10970930/main-putstrln.wasm.gz)


```hs
module Main where

import System.IO

main = do
  putStrLn "{\"name\":\"bob\"}"
  return 0
```


    wazero run main-putstrln.wasm

silently dies without printing anything

**Comment by evacchi** [comment]
ah that's right, the reason why it dies without any error it's because the `hFlush` is really the one complaining

**Comment by evacchi** [comment]
Unless I am horribly misguided (which I may be) this should fix the issue 😅 https://github.com/tetratelabs/wazero/pull/1239

**Comment by codefromthecrypt** [comment]
thanks for fixing this, @evacchi! we'll have a release candidate out today or tomorrow
